pkg/rollup/rollup.go

```diff
@@ -0,0 +1,133 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package rollup
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/jackc/pgx/v4"
+	"github.com/timescale/promscale/pkg/util"
+)
+
+type DownsampleResolution struct {
+	label      string
+	resolution util.DayDuration
+	retention  util.DayDuration
+}
+
+// Parse parses the rollup resolution string and returns true if it respects the format of downsampling resolution,
+// which is a string of comma separated label:resolution:retention.
+// Example: short:5m:90d,long:1h:395d
+func Parse(resolutionStr string) ([]DownsampleResolution, error) {
+	resolutions := strings.Split(resolutionStr, ",")
+	if len(resolutions) < 1 {
+		return nil, fmt.Errorf("resolutions cannot be less than 1")
+	}
+	var r []DownsampleResolution
+	for _, resolution := range resolutions {
+		resolution = strings.TrimSpace(resolution)
+		if resolution == "" {
+			continue
+		}
+
+		items := strings.Split(resolution, ":")
+		if len(items) != 3 {
+			return nil, fmt.Errorf("expected items not found: needed in format of label:resolution:retention but found %v for %s resolution", strings.Join(items, ":"), resolution)
+		}
+		lName := items[0]
+
+		var res util.DayDuration
+		err := res.UnmarshalText([]byte(items[1]))
+		if err != nil {
+			return nil, fmt.Errorf("error parsing resolution %s: %w", items[1], err)
+		}
+
+		var retention util.DayDuration
+		err = retention.UnmarshalText([]byte(items[2]))
+		if err != nil {
+			return nil, fmt.Errorf("error parsing retention %s: %w", items[2], err)
+		}
+		r = append(r, DownsampleResolution{lName, res, retention})
+	}
+	return r, nil
+}
+
+// EnsureRollupWith ensures "strictly" that the given new resolutions are applied in the database.
+// Note: It follows a "strict" behaviour, meaning any existing resolutions of downsampling in
```

Comment:
It would be good to describe that we want to preserve existing downsampling data if possible to avoid recalculations and only create the new items.

Reply:
This is not in line with the design. In the design discussion I recall that we wanted this "strict" behaviour.
We can do what you suggested (ideally I am for it), but for now, maybe we follow what's discussed in the design doc, for things to move quickly.